### PR TITLE
Implement CRC32 in pure OCaml

### LINKED ISSUE #165: undefined reference to `crc32_string' under Solo5
Building https://github.com/g2p/irc-unikernel/tree/fat-test-case fails with the ukvm target

```
_build/main.native.o : Dans la fonction « camlCrc__3 » :
lib/crc.ml:(.data+0xbdb48) : référence indéfinie vers « crc32_string »
lib/crc.ml:(.data+0xbdb50) : référence indéfinie vers « crc32_cstruct »
```

This is #127 with the Solo5 backend. Creating a new issue so it doesn't get lost.

### PR DESCRIPTION
**Pull request by g2p**
This fulfills the portability promise of ocaml-git:
a pure OCaml implementation of Git, portable to all OCaml platforms,
including Mirage, Solo5, and JavaScript.

The implementation is taken from Bitcask (MIT licensed).

Closes #165.

### CODE REVIEW

**Comment by hannesm** [comment]
It might be worth to put crc32 in a separate library (and have some test cases from wherever it is specified).

File: lib/git_crc32.mli [hunk 1]
```diff
@@ -0,0 +1,11 @@
+val cstruct: ?crc:int32 -> Cstruct.t -> int32
```
Comment (samoht) [inline lib/git_crc32.mli]: Do you mind adding a copyright header in the mli too and removing the extra lines at the end?

Comment (g2p) [inline lib/git_crc32.mli]: Done! Thank you for reviewing.

**Comment by samoht** [comment]
@g2p hum unfortunately, your changes seem to not be correct.

```
$ make
$ ./_build/lib_test/test.native test FS 10
[...]
+1033621us git.unix   [INFO] Writing /Users/thomas/git/ocaml-git/test-db/.git/refs/heads/test-fetch-1 (/Users/thomas/git/ocaml-git/test-db/.git/tmp/test-fetch-127c655write)
error: index CRC mismatch for object 348199320dc33614bc5d101b1e0e22eaea25b36b from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 12
error: index CRC mismatch for object 7e9fcafb80a4b5d306bb8efafe6bf694e28ce3dc from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 184
error: index CRC mismatch for object 51f8a0b3e50fd04e9b1d13202254c12749aaadf7 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 347
error: index CRC mismatch for object 4a11c6acbdb0d9fb5183f8ae86794b4cc1367d76 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 533
error: index CRC mismatch for object 85675f121da38f57dae4123195b77245b8dc7815 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 706
error: index CRC mismatch for object bf1320e33102831e4a2a951548709af1577094ed from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 887
error: index CRC mismatch for object 0163cf91efb58da39c9828635bd5cf6be5533e30 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 1043
error: index CRC mismatch for object 9936ab71a1c297a38de6485bc800a53303e515e5 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 1216
error: index CRC mismatch for object 9d453b3377f31bcacfe18ac71c0afbafa3defb8b from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 1394
error: index CRC mismatch for object 90bf66273cb045cd62d64ce4e78aba0825279861 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 1550
error: index CRC mismatch for object 7f1d3414f480881d22af4883567ce6e77a36cc9f from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 1710
error: index CRC mismatch for object d457c0aa028ab5c4091de116b4ea422338f6a945 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 1837
error: index CRC mismatch for object 5138ac2f84b8bd4e98dc3f6b0a29c9fc36006f2f from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 2238
error: index CRC mismatch for object 12470b11dcc0fa3602ead013c83966345f9fdd58 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 2362
error: index CRC mismatch for object 3639f14addb003984226f1a47b3c9adbfa9aa8b4 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 2414
error: index CRC mismatch for object ab0bcbfd9fc0c4d4423d84e8d635fefb8d2bb301 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 2729
error: index CRC mismatch for object 39b7ebb37e5d4593cec95291bd429aa868c445ce from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 3259
error: index CRC mismatch for object 9c2fb0a28afacd01f43380ea7da738175ce9261e from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 3545
error: index CRC mismatch for object fb33c47053b0d51f772a0d365209025727355221 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 3578
error: index CRC mismatch for object ed6a2121811249fe3a8cd72fc3d8e6c3334faf9b from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 3935
error: index CRC mismatch for object 6acfaeb953f7778c285f85f3e41ac0adfd6188fb from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 4027
error: index CRC mismatch for object a60718c0b1a9210edd7a9e6b7941badf2218008c from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 4280
error: index CRC mismatch for object c19966dd7971c03b34523557ace49220aa5c55d4 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 4327
error: index CRC mismatch for object 5c380d473b106dad66a1dec386472daf1d71b53c from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 89318
error: index CRC mismatch for object 5acf7a4873cf836c3c4352965696271b2cc0c051 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 94099
error: index CRC mismatch for object f93de30714d849b9812f9fb5554ca9a80b72f297 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 95910
error: index CRC mismatch for object 36f86304933d380046573480b1b2822289fc380c from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 96063
error: index CRC mismatch for object 6d08132e72a4b09dd3ad8766e2a7de340a022b36 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 96111
error: index CRC mismatch for object 7e83b75660cb4a276fa9ebf4150715ec1ef7f368 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 96956
error: index CRC mismatch for object b2829d9e929f2c0f4779dedc21f7610b41c30300 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 142908
error: index CRC mismatch for object 993d4cd8fadd8628a3712dca5b6f36e6c0eb79fd from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 143095
error: index CRC mismatch for object a8babcddc22eb2fda6350a2535326953b69c9130 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 143144
error: index CRC mismatch for object 9df08c14945bbe1850244639e74fd62a3ecc992c from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 143190
error: index CRC mismatch for object 22a37ef2d469997a361503cd0299c6a83273438f from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 143231
error: index CRC mismatch for object 289b3197aa5775554d6aa567bde2b8020880ef4b from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 143252
error: index CRC mismatch for object cd3e8e8042729b1d591109fdd8be3f9923248da3 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 143278
error: index CRC mismatch for object fc48d4b1303a1b4b53cd7bc5054f18d883ec8164 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 143358
error: index CRC mismatch for object 88e2ba0410bb30cf930ca7031d56ef821a74d8cb from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 143408
error: index CRC mismatch for object 52b02434d3aaa6fa6f3e186660c767049d022101 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 143727
error: index CRC mismatch for object bc7f8d7b070ef5414e0b7b486a180db855136786 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 144332
error: index CRC mismatch for object 7a7adc62716dfadac34972ba8a3878424acd6e27 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 144380
error: index CRC mismatch for object beb5b2f48b79a135e0b9c7596f6b541886447bcd from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 144591
error: index CRC mismatch for object 85e1730bdaac74543db01318fa09389cec75a177 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 144795
error: index CRC mismatch for object e36e86f788f5064cdb61d1e23a73e548d1d585a3 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 144902
error: index CRC mismatch for object 4c9d15c1fd9520ee3a032fc28717e444bb284603 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 145002
error: index CRC mismatch for object cf49f62593c9cc2c5f40e4f259083d6858a8d8d2 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 145047
error: index CRC mismatch for object ec826547b23971caaa9420bf142a857f639c42a0 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 145117
error: index CRC mismatch for object f2f673542b644dc7fec75527b8fadc17464237f0 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 145164
error: index CRC mismatch for object 9e8f78e52979d6cf613973ce3f38166796954867 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 145207
error: index CRC mismatch for object 018a9f7dec405c19b47d58e5bd667298df21f6c0 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 145257
error: index CRC mismatch for object 00204c121ea7c08efcb5290101fb6b58c827a0ee from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 145285
error: index CRC mismatch for object 412aaf90459e1cf78f95be3ae1699c95defea66a from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 145428
error: index CRC mismatch for object 12bb070edf022333dbe806055ec693267c0c7112 from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 145567
error: index CRC mismatch for object 5525f13c62bd0d9fb2023625b5b4e2161a2fbb5f from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 145586
error: index CRC mismatch for object cde3e89a6921b17f56281ed0360425add0e2d86b from .git/objects/pack/pack-24e9bbe8548ef5b94a2b3b32060da0a59d00c57d.pack at offset 145646
------------------------------------------------------------------------------------------
ASSERT fsck
------------------------------------------------------------------------------------------
[failure] Error fsck: expecting 0, got 4.
```

**Comment by hannesm** [comment]
did I say test cases? *scnr*

**Comment by g2p** [comment]
Sorry, fixed. I was expecting Travis to do all the work, should have figured out how to run the test suite.

**Pull request merged**